packages/task/src/node/task-server.ts

```diff
@@ -0,0 +1,232 @@
+import { FileUri } from "@theia/core/lib/node";
+import { TaskConfigurations, TaskConfigurationClient } from './task-configurations';
+
+@injectable()
+export class TaskServer implements ITaskServer, TaskConfigurationClient, Disposable {
+
+    /* Task client to send notifications to.  */
+    protected client: ITaskClient | undefined = undefined;
+    protected taskToDispose = new Map<number, DisposableCollection>();
+    protected readonly toDispose = new DisposableCollection();
+
+    protected workspaceRoot: string = '';
+
+    constructor(
+        @inject(ILogger) protected readonly logger: ILogger,
+        @inject(RawProcessFactory) protected readonly rawProcessFactory: RawProcessFactory,
+        @inject(TerminalProcessFactory) protected readonly terminalProcessFactory: TerminalProcessFactory,
+        @inject(TaskManager) protected readonly taskManager: TaskManager,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(TaskConfigurations) protected readonly taskConfigurations: TaskConfigurations
+    ) {
+        this.toDispose.push(taskConfigurations);
+
+        const taskManagerListener = taskManager.onDelete(id => {
+            const toDispose = this.taskToDispose.get(id);
+            if (toDispose !== undefined) {
+                toDispose.dispose();
+                this.taskToDispose.delete(id);
+            }
+        });
+        this.toDispose.push(taskManagerListener);
+        this.taskConfigurations.init(this);
+    }
+
+    async setWorkspaceRoot(workspaceRoot: string): Promise<void> {
+        this.workspaceRoot = workspaceRoot;
+        await this.taskConfigurations.setWorkspaceRoot(workspaceRoot);
+        return Promise.resolve();
+    }
+
+    dispose() {
+        this.client = undefined;
+        this.toDispose.dispose();
+    }
+
+    getTasks(): Promise<string[]> {
+        return Promise.resolve(this.taskConfigurations.getTaskLabels());
+    }
+
+    run(task: ITaskOptions | string): Promise<ITaskInfo> {
+        if (typeof task === 'string') {
+            return new Promise((resolve, reject) => {
+                this.taskConfigurations.getTask(task)
+                    .then(t => this.doRun(t).then(taskInfo => resolve(taskInfo)))
+                    .catch(e => {
+                        reject(new Error(e));
+                    });
+            });
+
+        } else {
+            return this.doRun(task);
+        }
+    }
+
+    protected doRun(options: ITaskOptions): Promise<ITaskInfo> {
+        let task: Task;
+        let proc: TerminalProcess | RawProcess;
+        const processOptions = (isWindows && options.windowsProcessOptions !== undefined) ?
+            options.windowsProcessOptions : options.processOptions;
+
+        const command = processOptions.command;
+
+        let cwd;
+        if (processOptions.options === undefined) {
+            processOptions.options = {
+                'cwd': '',
+                'env': ''
```

Comment:
We could use a proper processOptionsDefault here ?

Reply:
Can you explain this one a bit?